src/kudu/consensus/log_index.cc

```diff
@@ -193,23 +230,115 @@ Status LogIndex::OpenAllChunksOnStartup(Env *env) {
     VLOG(1) << "Opening index file on startup: " << fname << " for chunk idx " << chunk_idx;
 
     scoped_refptr<IndexChunk> chunk;
-    RETURN_NOT_OK(OpenAndInsertChunk(chunk_idx, &chunk));
+    RETURN_NOT_OK(OpenAndInsertChunk(chunk_idx, &chunk, /*should_mmap=*/false));
+  }
+
+  // mmap 'kNumChunksToMmap' chunks. Note that the latest chunks are mmaped
+  // (chunks having the highest chunk_idx)
+  int64_t mmapped_chunks = 0;
+  for (auto rit = open_chunks_.rbegin(); rit != open_chunks_.rend(); ++rit) {
+    if (mmapped_chunks == kNumChunksToMmap) {
+      break;
+    }
+
+    RETURN_NOT_OK(rit->second->Mmap());
+    mmapped_chunks++;
+  }
+
+  return Status::OK();
+}
+
+void LogIndex::SetNumMmapChunks(int64_t num_chunks) {
+  if (num_chunks <= 0)
+    return;
+
+  std::lock_guard<simple_spinlock> l(open_chunks_lock_);
+  kNumChunksToMmap = num_chunks;
+
+  // If necessary, unmap additional chunks
+  if (open_chunks_.size() <= kNumChunksToMmap) {
+    return;
+  }
+
+  // get the number of chunks that are currently mmapped
+  int64_t num_chunks_mmapped = 0;
+  for (auto it = open_chunks_.begin(); it != open_chunks_.end(); ++it) {
+    if (it->second->IsMmapped()) {
+      num_chunks_mmapped++;
+    }
   }
 
+  // unmap additional chunks (starting with the oldest chunks)
+  for (auto it = open_chunks_.begin();
+      it != open_chunks_.end() && num_chunks_mmapped > kNumChunksToMmap;
+      ++it) {
+    if (it->second->IsMmapped()) {
+      it->second->Munmap();
+      num_chunks_mmapped--;
+    }
+  }
+}
+
+void LogIndex::SetNumEntriesPerChunkForTest(int64_t entries) {
+  // WARNING: This should be called only for tests. Changing numer of entries
+  // per chunk is dangerous and could render the chunk files unreadable
+  kEntriesPerIndexChunk = entries;
+}
+
+Status LogIndex::MmapChunk(scoped_refptr<IndexChunk> *chunk) {
+  if (open_chunks_.size() < kNumChunksToMmap) {
+    RETURN_NOT_OK((*chunk)->Mmap());
+    return Status::OK();
+  }
+
+  // The victim that needs to be unmapped is the oldest chunk (i.e the chunk
+  // with the oldest chunk_idx). See documentation in log_index.h for more
+  // details.
+  //
+  // Note that we have to reverse iterate through the open_chunks_ while the
+  // caller is holding onto the open_chunks_lock_. With 'open_chunks_' map
+  // having aonly a few hundred entries, this should be acceptable (to keep
+  // things simple)
+  int64_t num_chunks_mmapped = 0;
+  auto rit = open_chunks_.rbegin();
+  for (; rit != open_chunks_.rend(); ++rit) {
+    if (rit->second->IsMmapped()) {
+      num_chunks_mmapped++;
+    }
+
+    if (num_chunks_mmapped == kNumChunksToMmap)
+      break;
+  }
+
+  // If there are 'kNumChunksToMMap' chunks already mmaped, then unmap the
+  // 'victim' chunk
+  if (num_chunks_mmapped == kNumChunksToMmap && rit != open_chunks_.rend()) {
+    rit->second->Munmap();
```

Comment:
since you have added a mutex, you can maintain a simple LRU list here.
mmaped_chunks_
After Mmap() below you add it to linked list.
And when you Munmap here, you can find the head of the list and unmap it.

Reply:
Thought about it too, but it makes `LogIndex::GC()` more complicated. `LogIndex::GC()` is written in such a way that it does not have to hold onto `open_chunks_lock_` when unlinking the file (that is doing disk IO) - with two different maps this can introduce races or one needs to add more locks and refactor that part. So, I decided to keep this simple for now.